### PR TITLE
Spw grid view

### PR DESCRIPTION
**Pull request by will-moore**
# What this PR does

First PR towards support for new SPW 'grid' layout.
This introduces a few changes, but not "full support" of https://github.com/openmicroscopy/design/issues/52
- When a Well is selected, right panel shows Well details & annotations (instead of Image)
- Image thumbnails are shown for selected well(s) below plate.
- Click to select an Image - this image will be shown in right panel.
- Plate and Well Images can both be zoomed with sliders

This PR does not support a number of 'grid view' features (to come in future PRs)
- Spacial layout of fields within well
- Set number of columns in grid-view

screen shot 2016-09-23 at 15 43 35
# Testing this PR
- Test Well annotations (Wells have not previously been annotatable)
  - Add, remove etc all types of annotations on Well.
  - Batch select Wells and add/remove annotations.
  - Metadata and Preview panels should be disabled for Wells
- Test loading of thumbnails for selected wells:
  - Select wells by Click or multi-select by drag or shift-click
  - For each selected well, a row of images will be loaded in bottom pane
  - Bottom pane can be resized by drag handle
  - Plate and well Images can be zoomed with sliders in bottom-right of each panel
- Clicking a single Image in bottom panel will load that image in Right panel.
- Can select Images / Wells in turn.
# TODO for this PR:
- [x] Multi-select Images
- [x] Indicate selected images with border etc.
- [x] Preserve zoom sizes when switching between Plates
- [x] Refresh Well grid thumbnails ( 'Save To All' or just 'Save')
- [ ] Handle browsing Tag tree -> Well. What do we show in tree/centre/right panels?
- [x] Tests for `marshal_tagged` in tree.py to test `wells`

### CODE REVIEW

**Comment by gusferguson** [comment]
Tested using http://web-dev-merge.openmicroscopy.org/webclient/usertags/?show=tag-14953 user-3 Safari

Works as expected:
- annotation of individual wells
- batch annotation of wells
- selection of wells by dragging or shift-click
- zooming of top and bottom panes - plate and well Images
- right pane displaying correct information

Not working as expected:
- tagged wells are not showing up in the Tags tab tree

**Comment by dominikl** [comment]
I still struggle with this whole data model... in the clients we have runs and fields, in the data model there are plateacquisitions, wellsamples, etc. is that the same, how does that fit together?

**Comment by mtbc** [comment]
Yes, plate acquisitions are runs, well samples are fields. You'll tend to find the former terms more in the guts of the system and the latter more as we face users.

**Comment by will-moore** [comment]
As expected, lots of robot failures: https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/400/robot/

**Comment by will-moore** [comment]
To test last commit: resize thumbnails in both panels and also resize lower panel.
Then load a different Plate - thumb sizes and panel height should be maintained.

**Comment by pwalczysko** [comment]
Ditto to all points raised by @gusferguson in https://github.com/openmicroscopy/openmicroscopy/pull/4863#issuecomment-249811137,  **including** the Bug with Tags not containing wells in Tags tab.

I have one more issue: When the numerals are added (going from 9 to 10, i.e. from 1 to 2 numerals) there is a sudden formatting shift in lower pane (images). I do not see this as desirable.
screen shot 2016-09-27 at 11 06 40

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#458](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/458/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/458/consoleText) for more details.
Possible conflicts:
  - PR #4708 will-moore 'Webgateway api2'
    - components/tools/OmeroWeb/omeroweb/webclient/views.py
  - PR #4839 will-moore 'Codomain web'
    - components/tools/OmeroWeb/omeroweb/webclient/views.py
  - PR #4842 will-moore 'Tag removal performance'
    - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by will-moore** [comment]
Few ```test_marshal_tagged``` failing tests: https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-integration-Python27/338/testReport/

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#467](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/467/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/467/consoleText) for more details.
Possible conflicts:
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4839 will-moore 'Codomain web'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4842 will-moore 'Tag removal performance'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- Upstream changes
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#468](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/468/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/468/consoleText) for more details.
Possible conflicts:
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4839 will-moore 'Codomain web'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#470](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/470/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/470/consoleText) for more details.
Possible conflicts:
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4839 will-moore 'Codomain web'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by pwalczysko** [comment]
Thanks a lot for this PR being in the build now.

All seems to work well, just one issue with formatting:

When there are a bit more fields (48) and couple of rows in the bottom panel, the zoom-in-slider is overlayed on top of the fields thumbs, see screenshot below.
screen shot 2016-10-11 at 11 41 11

**Comment by pwalczysko** [comment]
RFE: If I have a little bit more fields (700), then in the bottom panel I can get quickly to the situation that I do not know which well I am on (see screenshot below). In Insight, I will get a number above the thumb as well as a tooltip on the field in bottom panel.

screen shot 2016-10-11 at 11 46 42

Insight:

screen shot 2016-10-11 at 12 28 42

**Comment by will-moore** [comment]
I think we need to decide for web and #4874 how many thumbnails we should load at once (maybe do this in follow-up PRs?)
E.g. If I select a single well, we may show up to 50 thumbnails for that well, but if I select 20 wells, we should probably only show 5 thumbnails per well.
We could then allow to "Load more" for a single Well (E.g. 50 more) OR to paginate all selected (E.g. next 5 thumbs for all 20 wells)?

**Comment by will-moore** [comment]
Added image name as tool-tip to images:

screen shot 2016-10-19 at 14 48 29

**Comment by will-moore** [comment]
Improve wrapping of images onto multiple rows - all indented same now:

screen shot 2016-10-19 at 15 50 56

**Comment by gusferguson** [comment]
@will-moore 

Tested with http://web-dev-merge.openmicroscopy.org/webclient/ user 4 Safari

The alignment is working as expected.
One problem I found was that when there are lots of fields in multi-selected wells it is difficult to tell where the one cell ends and the next begins in the grid view - see screenshot.
Could we perhaps make the deciding line darker grey and maybe align the well identifier to the top of the row in the grid view so that if there are so many fields that the row wraps then it is clear where the one well fields end and the next begins.

webclient

**Comment by will-moore** [comment]
@gusferguson I've improved the separation between wells. Do you think we also need the label aligned to the top or is this sufficient? (I kinda like it aligned to the middle)?

screen shot 2016-10-20 at 11 05 26

File: components/tools/OmeroWeb/omeroweb/webclient/static/webclient/javascript/ome.tree.js [hunk 1]
```diff
@@ -631,14 +631,22 @@ $(function() {
                                 });
                             }
 
-                            // Add plates to the jstree data structure
+                            // Add acquistions (runs) to the jstree data structure
```
Comment (jburel) [inline components/tools/OmeroWeb/omeroweb/webclient/static/webclient/javascript/ome.tree.js]: typo

File: components/tools/OmeroWeb/test/integration/test_tree.py [hunk 2]
```diff
@@ -1174,7 +1219,7 @@ def tagset_hierarchy_userA_groupA(request, userA, userB,
     image_link.child = TagAnnotationI(tags[0].id.val, False)
     get_connection(userB).getUpdateService().saveObject(image_link)
 
-    # links is: project, dataset, image, screen, plate, acquisition
+    # links is: project, dataset, image, screen, plate, acquisition, well, well
```
Comment (jburel) [inline components/tools/OmeroWeb/test/integration/test_tree.py:1222]: duplicate

Comment (will-moore) [reply components/tools/OmeroWeb/test/integration/test_tree.py:1222]: That's what's returned in the array (2 wells).

Comment (jburel) [reply components/tools/OmeroWeb/test/integration/test_tree.py:1222]: why is that?

Comment (will-moore) [reply components/tools/OmeroWeb/test/integration/test_tree.py:1222]: Because we want to test the retrieval of 2 wells

**Comment by will-moore** [comment]
@pwalczysko Last commits fix the slider display as reported in https://github.com/openmicroscopy/openmicroscopy/pull/4863#issuecomment-252878080

screen shot 2016-10-21 at 15 28 23

File: components/tools/OmeroWeb/test/integration/test_tree.py (hunk 2 above)
Comment (jburel) [reply components/tools/OmeroWeb/test/integration/test_tree.py:1222]: ok

**Comment by pwalczysko** [comment]
The slider display and the reformatting of the borders is great. 
Happy to merge, but probably one more look of @gusferguson and or @bramalingam would be good.

**Comment by gusferguson** [comment]
Agree with @pwalczysko - layout and sliders look good.
Good to merge from my POV.

**Comment by will-moore** [comment]
Great, if we can get this merged now, I can fix conflicts with #4708 and re-open it. Thanks.

**Comment by jburel** [comment]
One item that has not been addressed in this PR is the unticked item in the description
i.e. what to display when a well is tagged and display.

We can certainly do that in another PR

Currently we have for example B4, and we have no way to find where that well is
e.g. user-1 test tag
when an image is tagged at least we know where it is located

**Comment by jburel** [comment]
https://trello.com/c/GxqDnD0P/211-spw-follow-up for next actions

File: components/tools/OmeroWeb/test/integration/test_tree.py (hunk 2 above)
Comment (jburel) [reply components/tools/OmeroWeb/test/integration/test_tree.py:1222]: could you add a check for naming of columns/rows i.e. number vs letter

Comment (will-moore) [reply components/tools/OmeroWeb/test/integration/test_tree.py:1222]: That's what we're testing here: See `def expected_wells(user, wells):` which generates the dictionary that we compare to what's returned by the tagged json, and includes 'name'.

Comment (jburel) [reply components/tools/OmeroWeb/test/integration/test_tree.py:1222]: thanks I missed it

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#489](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/489/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/489/consoleText) for more details.
Possible conflicts:
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- Upstream changes
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#490](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/490/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/490/consoleText) for more details.
Possible conflicts:
- PR #4703 will-moore 'histograms'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- Upstream changes
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#491](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/491/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/491/consoleText) for more details.
Possible conflicts:
- PR #4703 will-moore 'histograms'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- Upstream changes
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#492](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/492/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/492/consoleText) for more details.
Possible conflicts:
- PR #4703 will-moore 'histograms'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- Upstream changes
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#493](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/493/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/493/consoleText) for more details.
Possible conflicts:
- PR #4703 will-moore 'histograms'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- Upstream changes
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#494](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/494/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/494/consoleText) for more details.
Possible conflicts:
- PR #4703 will-moore 'histograms'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- PR #4708 will-moore 'Webgateway api2'
  - components/tools/OmeroWeb/omeroweb/webclient/views.py
- Upstream changes
  - components/tools/OmeroWeb/omeroweb/webclient/views.py

**Comment by jburel** [comment]
Merging so we can move forward

**Pull request merged**